### PR TITLE
Auto Enable/Disable KVM hosts

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR allows CloudStack to auto-disable and auto-enable KVM hosts resource state based on customisable host/hypervisor health checks.

#### Management server changes

A new cluster configuration 'enable.kvm.host.auto.enable.disable' is added, disabled by default.

This PR extends the AgentManager to process StartupRoutingCommand and PingRoutingCommand with the host health check result:
 - If the host health check result is null, do nothing.
 - If the host health check result is true, enable the host resource state if it is Disabled.
 - If the host health check result is false, disable the host resource state if it is Enabled.

On every automatic enable or disable event, the management server will send an alert to the admin and add/update an automatic annotation (comment) on the specific host.

- If a host gets auto-disabled by a health check failure, then it can be auto-enabled when the health check succeeds. But if the host gets disabled by the admin, then it must not be auto-enabled when the health check succeeds (manual host disabling takes precedence over the auto-enabling of a host).
- If a host gets auto-disabled by a health check failure, the admin could enable the host but unless they also disable the health check on the host then it will just get disabled again when the health check fails

CloudStack controls when a host can/cannot be auto-enabled or auto-disabled by a host detail record (on the host_details table) with key ‘autoenablekvmhost’, in the following way:
- If a host is auto-enabled or auto-disabled and there is no host detail with key ‘autoenablekvmhost’ for that host, then a new host detail record is created with the key ‘autoenablekvmhost’ and is set to ‘true’. (indicating that the host can be auto-enabled/auto-disabled)
- If the administrator manually disables a host and there is a host detail with key ‘autoenablekvmhost’ for that host, then the host detail ‘autoenablekvmhost’ is set to ‘false’ (indicating that the host cannot be auto-enabled if the health check succeeds)
- If the administrator manually enables a host and there is a host detail with key ‘autoenablekvmhost’ for that host, then the host detail ‘autoenablekvmhost’ is set to ‘true’ (indicating that the host can be auto-disabled if the health check fails)
- If the cluster setting is disabled, and the administrator enables/disables hosts in the cluster, then the host detail with key ‘autoenablekvmhost’ is not created for the hosts. (preserving the usual behavior)

#### KVM host Health Check

Set the path of an executable file/script for host health check on the property 'agent.health.check.script.path' on `agent.properties` 

- The health check result is true, if the script is executed successfully and the exit code is 0
- The health check result is false, if the script is executed successfully and the exit code is 1
- The health check result is null, if 
   - Script file is not specified, or
   - Script file does not exist, or 
   - Script file is not accessible by the user of the cloudstack-agent process, or
   - Script file is not executable
   - There are errors when the script is executed (exit codes other than 0 or 1)

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
2xRocky8 KVM hosts in 1 cluster -> put host1 into maintenance, test host2
- Enable the cluster setting wihout providing a health script to the host -> no automatic enable/disable
- Enable the cluster setting, set a script that doesn't have permission rights -> no automatic enable/disable
- Enable the cluster setting, set a script that succeeds (ping the management server) -> observe the health check result is successful, but since the host is Enabled is not transitioned to any state
- Enable the cluster setting, set a script that fails (ping an unreacheable host) -> observe the health check result is false, the host2 is put into Disabled state
- Enable the host2 manually -> the host2 is set to Enabled state
- The next failure sets the hosts2 into Disabled state again
- Modify the script again so it succeeds, the host2 goes into Enabled state, then disable manually the host2 -> the future successful health checks do not put the host into Disabled state

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: test/integration/smoke/test_host_control_state.py [hunk 1]
```diff
@@ -250,3 +255,234 @@ def test_router_host_control_state(self):
...
+        cls._cleanup = [
+        ]
+        return
+
+    @classmethod
+    def tearDownClass(cls):
+        super(TestAutoEnableDisableHost, cls).tearDownClass()
+
+    def setUp(self):
+        self.apiclient = self.testClient.getApiClient()
```
Comment (slavkap) [inline test/integration/smoke/test_host_control_state.py:292]: Probably you have to remove the whole implementation of the setUp() method. You already have the `apiclient` from the `setUpClass()` and the `self.cleanup` isn't used

File: test/integration/smoke/test_host_control_state.py [hunk 2]
```diff
@@ -250,3 +255,234 @@ def test_router_host_control_state(self):
...
+
+        cls.template = get_template(
+            cls.apiclient,
+            cls.zone.id,
+            cls.hypervisor
+        )
+
+        cls.logger = logging.getLogger('TestAutoEnableDisableHost')
+
+        cls._cleanup = [
```
Comment (slavkap) [inline test/integration/smoke/test_host_control_state.py:283]: The _cleanup isn't used also

File: test/integration/smoke/test_host_control_state.py [hunk 3]
```diff
@@ -250,3 +255,234 @@ def test_router_host_control_state(self):
...
+        cls.services = cls.testClient.getParsedTestDataConfig()
+        # Get Zone, Domain and templates
+        cls.zone = get_zone(cls.apiclient, cls.testClient.getZoneForTests())
+        cls.hypervisor = cls.testClient.getHypervisorInfo()
+        cls.hostConfig = cls.config.__dict__["zones"][0].__dict__["pods"][0].__dict__["clusters"][0].__dict__["hosts"][0].__dict__
+        if cls.hypervisor.lower() not in ['kvm']:
+            cls.hypervisorNotSupported = True
+            return
+
+        cls.template = get_template(
```
Comment (slavkap) [inline test/integration/smoke/test_host_control_state.py:275]: The template isn't used

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 4]
```diff
@@ -1249,6 +1250,52 @@ public AgentHandler(final Task.Type type, final Link link, final byte[] data) {
...
+                s_logger.error(String.format("Unable to find host with ID: %s", hostId));
+                return;
+            }
+            if (!BooleanUtils.toBoolean(EnableKVMAutoEnableDisable.valueIn(host.getClusterId()))) {
+                s_logger.debug(String.format("%s is disabled for the cluster %s, cannot process the health check result " +
+                        "received for the host %s", EnableKVMAutoEnableDisable.key(), host.getClusterId(), host.getName()));
+                return;
+            }
+
+            String allocationState = hostHealthCheckResult ? "Enable" : "Disable";
```
Comment (weizhouapache) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:1268]: can define a enum ?

File: server/src/main/java/com/cloud/resource/ResourceManagerImpl.java [hunk 5]
```diff
@@ -1774,73 +1780,142 @@ public boolean checkAndMaintain(final long hostId) {
...
+        ResourceState.Event resourceEvent = getResourceEventFromAllocationStateString(allocationState);
+        DetailVO hostDetail = _hostDetailsDao.findDetail(host.getId(), ApiConstants.AUTO_ENABLE_KVM_HOST);
+
+        if ((host.getResourceState() == ResourceState.Enabled && resourceEvent == ResourceState.Event.Enable) ||
+                (host.getResourceState() == ResourceState.Disabled && resourceEvent == ResourceState.Event.Disable)) {
+            s_logger.info(String.format("The host %s is already on the allocated state", host.getName()));
+            return false;
+        }
+
+        if (autoEnableDisableKVMSetting && isUpdateFromHostHealthCheck && hostDetail != null &&
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/resource/ResourceManagerImpl.java:1827]: can you extract as a method ?

**Review by weizhouapache — APPROVED** [review]
overall code lgtm

**Comment by nvazquez** [comment]
Thanks @slavkap @weizhouapache, I've addressed your reviews
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test matrix

**Comment by borisstoyanov** [comment]
@blueorangutan test vmware7

**Comment by borisstoyanov** [comment]
@blueorangutan help

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-70u3

**Review by borisstoyanov — APPROVED** [review]
LGTM, I've manually tested the enablement and features of this PR

**Review by andrijapanicsb — CHANGES_REQUESTED** [review]

File: test/integration/smoke/test_host_control_state.py [hunk 6]
```diff
@@ -250,3 +255,220 @@ def test_router_host_control_state(self):
...
+    @attr(tags=["basic", "advanced"], required_hardware="false")
+    @skipTestIf("hypervisorNotSupported")
+    def test_03_enable_host_does_not_override_auto_disable_kvm_host(self):
+        """Test to override the auto-disabling of a KVM host by an administrator
+
+        # Validate the following:
+        # 1. Enable the KVM Auto Enable/Disable Feature
+        # 2. Set a health check script that fails and observe the host is Disabled
+        # 3. Make the host Enabled
+        # 4. Verify the host does not get auto-disabled after the previous step
```
[the remark this answers is not in the document]
Comment (nvazquez) [reply test/integration/smoke/test_host_control_state.py]: Thanks @andrijapanicsb - fixed

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Review by andrijapanicsb — APPROVED** [review]

**Comment by andrijapanicsb** [comment]
LGTM

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
@nvazquez @borisstoyanov - is this unit test and marvin test coverage decent? (I'm not sure if codecov is right - if it isn't, can you check and advise?)

**Pull request merged**